### PR TITLE
[query] more lapack bindings, support using mkl

### PR DESCRIPTION
**Pull request by patrick-schultz**
Adds support for using mkl, which is needed for some of the added lapack methods.

To use, install mkl, then set the environment variable `HAIL_MKL_PATH` to the directory containing the mkl dylib files. In particular, it should contain `libmkl_rt.dylib`.

On my laptop, I installed mkl through conda with `conda install -c intel mkl`, then set `HAIL_MKL_PATH=~/opt/miniconda3/envs/hail/lib/`.

### CODE REVIEW

File: hail/src/main/scala/is/hail/linalg/LAPACK.scala [hunk 1]
```diff
@@ -20,19 +21,25 @@ class UnderscoreFunctionMapper extends FunctionMapper {
 object LAPACK {
   private[this] val libraryInstance = ThreadLocal.withInitial(new Supplier[LAPACKLibrary]() {
     def get() = {
-      val standard = Native.loadLibrary("lapack", classOf[LAPACKLibrary]).asInstanceOf[LAPACKLibrary]
+      val library_name = if (HailContext.isInitialized && HailContext.getFlag("mkl_path") != null) {
+        NativeLibrary.addSearchPath("mkl_rt", "/Users/pschulz/opt/miniconda3/envs/hail/lib")
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/linalg/LAPACK.scala:25]: You've hardcoded a path to your own laptop here.

**Review by johnc1231 — CHANGES_REQUESTED** [review]

File: hail/src/main/scala/is/hail/linalg/LAPACK.scala (hunk 1 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/linalg/LAPACK.scala:25]: Whoops, thanks. I had that using the environment variable, then I lost that code and had to quickly reconstruct it.

**Comment by patrick-schultz** [comment]
My hacky attempt to allow setting a path to an alternate lapack library clearly doesn't work on the cluster: it's trying to look up a flag on the HailContext on workers. Open to suggestions on the right design here.

**Comment by johnc1231** [comment]
Hmmm...  I guess we could just have it look for an environment variable, instead of asking for a flag on `HailContext`. Easy enough to set the environment variable in the `hailctl` init scripts. 

You should maybe name it `HAIL_MKL_PATH` instead of `MKL_PATH`, so people know what this variable is for / doesn't conflict with other tools variables.

**Comment by patrick-schultz** [comment]
Oh, good idea

**Comment by patrick-schultz** [comment]
Changed to use the environment variable directly. I also updated the PR description with details on how to install mkl and make Hail load it.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala [hunk 2]
```diff
@@ -244,6 +245,65 @@ object SNDArray {
     }
   }
 
+  def copy(cb: EmitCodeBuilder, _X: SNDArrayCode, _Y: SNDArrayCode): Unit = {
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:248]: I feel like something about the name of this function should suggest that it is only for copying vectors.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala [hunk 3]
```diff
@@ -244,6 +245,65 @@ object SNDArray {
...
+  }
+
+  def gemv(cb: EmitCodeBuilder, trans: String, A: SNDArrayCode, X: SNDArrayCode, Y: SNDArrayCode): Unit = {
+    gemv(cb, trans, 1.0, A, X, 1.0, Y)
+  }
+
+  def gemv(cb: EmitCodeBuilder, trans: String, alpha: Code[Double], _A: SNDArrayCode, _X: SNDArrayCode, beta: Code[Double], _Y: SNDArrayCode): Unit = {
+    val A = _A.memoize(cb, "copy_A")
+    val X = _X.memoize(cb, "copy_X")
+    val Y = _Y.memoize(cb, "copy_Y")
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:283]: Presumably `gemv_A`, `gemv_X`, `gemv_Y` were what you were going for?

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala [hunk 4]
```diff
@@ -244,6 +245,65 @@ object SNDArray {
...
+
+    val Seq(m, n) = A.shapes(cb)
+    if (trans == "N")
+      cb.ifx(X.shapes(cb)(0).cne(n) || Y.shapes(cb)(0).cne(m), cb._fatal("gemv: incompatible dimensions"))
+    else
+      cb.ifx(X.shapes(cb)(0).cne(m) || Y.shapes(cb)(0).cne(n), cb._fatal("gemv: incompatible dimensions"))
+
+    val ldA = (A.strides(cb)(1).toI >> 3).max(1)
+    val ldX = (X.strides(cb)(0).toI >> 3).max(1)
+    val ldY = (Y.strides(cb)(0).toI >> 3).max(1)
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:297]: Don't have to do anything with this comment, but I wonder if we shouldn't just have a `byteStride(i)` function on `SNDArrayValue` that looks up the stride at index `i` and does the bitshift?

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala [hunk 5]
```diff
@@ -273,6 +333,159 @@ object SNDArray {
       C.firstDataAddress(cb), ldC)
   }
 
+  def trmm(cb: EmitCodeBuilder, side: String, uplo: String, transA: String, diag: String,
+    alpha: Code[Double], _A: SNDArrayCode, _B: SNDArrayCode): Unit = {
+    val A = _A.memoize(cb, "copy_A")
+    val B = _B.memoize(cb, "copy_B")
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:339]: `trmm_A` and `trmm_B`?

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala [hunk 6]
```diff
@@ -244,6 +245,65 @@ object SNDArray {
     }
   }
 
+  def copy(cb: EmitCodeBuilder, _X: SNDArrayCode, _Y: SNDArrayCode): Unit = {
+    val X = _X.memoize(cb, "copy_X")
+    val Y = _Y.memoize(cb, "copy_Y")
+    assertVector(X, Y)
+    val n = X.shapes(cb)(0)
+    cb.ifx(Y.shapes(cb)(0).cne(n), cb._fatal("copy: vectors have different sizes"))
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:253]: If you include the sizes in the message your life will be easier in the future.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala [hunk 7]
```diff
@@ -273,6 +333,159 @@ object SNDArray {
...
+    alpha: Code[Double], _A: SNDArrayCode, _B: SNDArrayCode): Unit = {
+    val A = _A.memoize(cb, "copy_A")
+    val B = _B.memoize(cb, "copy_B")
+    assertMatrix(A, B)
+    assertColMajor(cb, A, B)
+
+    val Seq(m, n) = B.shapes(cb)
+    val Seq(a0, a1) = A.shapes(cb)
+    cb.ifx(a1.cne(if (side == "left") m else n), cb._fatal("trmm: incompatible matrix dimensions"))
+    cb.ifx(a0 < a1, cb._fatal("trmm: A has fewer rows than cols"))
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:346]: If you include the row and col count in the message your life will be easier in the future.

**Review by johnc1231 — CHANGES_REQUESTED** [review]

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala (hunk 2 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:248]: Good point. I was just copying blas/lapack names (dropping the first element type letter), but I can make this more descriptive.

File: hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala (hunk 4 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala:297]: Yeah, I like that. In fact, I'd like it if the SNDArray interface completely abstracted over whether strides are stored as bytes or elements.

**Review by johnc1231 — APPROVED** [review]

**Pull request merged**